lib/index.js

```diff
@@ -477,47 +487,50 @@ Socket.prototype.send = function(msg, flags) {
 // This helper is called from `send` above, and in response to
 // the watcher noticing the signaller fd is readable.
 Socket.prototype._flush = function() {
-  var args;
+  var flags, args, emitArgs;
 
   // Don't allow recursive flush invocation as it can lead to stack
   // exhaustion and write starvation
   if (this._flushing) return;
 
-  if(zmq.STATE_CLOSED == this._zmq.state) return;
-
   this._flushing = true;
   while (true) {
     // If an Async thread that may call a zmq function is in flight
     // do not call any zmq functions.
-    if (this._zmq.state != 0) {
+    if (this._zmq.state !== zmq.STATE_READY) {
       return;
     }
 
-    var emitArgs
-      , flags = this._ioevents;
+    flags = this._zmq.getsockopt(zmq.ZMQ_EVENTS);
 
     if (!this._outgoing.length) flags &= ~zmq.ZMQ_POLLOUT;
-    if (!flags) break;
+    if (!flags) return;
 
     if (flags & zmq.ZMQ_POLLIN) {
         emitArgs = ['message'];
 
```

Comment:
I'm not sure if I fully understand what you mean by "a recv() function that flushes", but I was gonna make a 2nd PR (splitting up sending and receiving). The reason why I'm hesitating to send that in yet, is that I kind of want to go baby steps here. The slightest screw up will cause regressions, and I want to avoid that at all costs.